rfc/text/advanced/ap_authentication_scram.md

```diff
@@ -0,0 +1,387 @@
+
+
+##### Authentication Exchange
+
+WAMP-SCRAM uses a single round of challenge/response pairs after the client authentication request and before the authentication outcome.
+
+The mapping of RFC5802 messages to WAMP messages is as follows:
+
+SCRAM Message                              | WAMP Message
+----------------------                     | ------------
+`client-first-message`                     | `HELLO`
+`server-first-message`                     | `CHALLENGE`
+`client-final-message`                     | `AUTHENTICATE`
+`server-final-message` with `verifier`     | `WELCOME`
+`server-final-message` with `server-error` | `ABORT`
+
+
+###### Initial Client Authentication Message
+
+WAMP-SCRAM authentication begins with the client sending a `HELLO` message specifying the `wamp-scram` method as (one of) the authentication methods:
+
+{align="left"}
+```javascript
+    [1, "realm1",
+        {
+            "roles": ...,
+            "authmethods": ["wamp-scram"],
+            "authid": "user",
+            "authextra":
+                {
+                    "nonce": "egVDf3DMJh0=",
+                    "channel_binding": null
+                }
+
+        }
+    ]
+```
+where:
+
+1. `authid|string`: The identity of the user performing authentication. This corresponds to the `username` parameter in RFC5802.
+2. `authextra.nonce|string`: A base64-encoded sequence of random octets, generated by the client. See [Nonces](#nonces).
+3. `authextra.channel_binding|string`: Optional string containing the desired channel binding type. See [Channel Bindings](#channel-bindings).
+
+Upon receiving the `HELLO` message, the server MUST terminate the authentication process by sending an `ABORT` message under any of the following circumstances:
+
+* The server does not support the WAMP-SCRAM `authmethods`, and there are no other methods left that the server supports for this `authid`.
+* The the server does not support the requested `channel_binding` type.
+* (Optional) The server does not recognize the given `authid`. In this case, the server MAY proceed with a mock `CHALLENGE` message to avoid leaking information on the existence of usernames. This mock `CHALLENGE` SHOULD contain a generated `salt` value that is always the same for a given `authid`, otherwise an attacker may discover that the user doesn't actually exist.
+
+###### Initial Server Authentication Message
+
+If none of the above failure conditions apply, and the server is ready to let the client authenticate using WAMP-SCRAM, then it SHALL send a `CHALLENGE` message:
+
+{align="left"}
+```javascript
+    [4, "wamp-scram",
+        {
+            "nonce": "egVDf3DMJh0=SBmkFIh7sSo=",
+            "salt": "aBc+fx0NAVA=",
+            "kdf": "pbkdf2",
+            "iterations": 4096,
+            "memory": null
```

Comment:
Maybe just name this "cost" so it can be consistent between `pbkdf2` and `argon2id`?

Reply:
Both `pbkdf2` and `argon2id` refer to this parameter as "iterations". It was bcrypt that was different, but we've now agreed to remove bcrypt as a possible KDF.

Reply:
ah, okay then nevermind :)